drivers/stm32/canard_stm32.c

```diff
@@ -0,0 +1,526 @@
+    const bool rtr_b = b & CANARD_CAN_FRAME_RTR;
+    if ((clean_a == clean_b) && (rtr_a != rtr_b))
+    {
+        return rtr_b;
+    }
+
+    /*
+     * Plain ID arbitration - greater value loses.
+     */
+    return clean_a < clean_b;
+}
+
+/// Converts libcanard ID value into the bxCAN TX ID register format.
+static uint32_t convertFrameIDCanardToRegister(const uint32_t id)
+{
+    uint32_t out = 0;
+
+    if (id & CANARD_CAN_FRAME_EFF)
+    {
+        out = ((id & CANARD_CAN_EXT_ID_MASK) << 3) | CANARD_STM32_CAN_TIR_IDE;
+    }
+    else
+    {
+        out = ((id & CANARD_CAN_STD_ID_MASK) << 21);
+    }
+
+    if (id & CANARD_CAN_FRAME_RTR)
+    {
+        out |= CANARD_STM32_CAN_TIR_RTR;
+    }
+
+    return out;
+}
+
+/// Converts bxCAN TX/RX (sic! both RX/TX are supported) ID register value into the libcanard ID format.
+static uint32_t convertFrameIDRegisterToCanard(const uint32_t id)
+{
+#if (CANARD_STM32_CAN_TIR_RTR != CANARD_STM32_CAN_RIR_RTR) ||\
+    (CANARD_STM32_CAN_TIR_IDE != CANARD_STM32_CAN_RIR_IDE)
+# error "RIR bits do not match TIR bits, TIR --> libcanard conversion is not possible"
+#endif
+
+    uint32_t out = 0;
+
+    if ((id & CANARD_STM32_CAN_RIR_IDE) == 0)
+    {
+        out = (CANARD_CAN_STD_ID_MASK & (id >> 21));
+    }
+    else
+    {
+        out = (CANARD_CAN_EXT_ID_MASK & (id >> 3)) | CANARD_CAN_FRAME_EFF;
+    }
+
+    if ((id & CANARD_STM32_CAN_RIR_RTR) != 0)
+    {
+        out |= CANARD_CAN_FRAME_RTR;
+    }
+
+    return out;
+}
+
+
+static bool waitMSRINAKBitStateChange(volatile const CanardSTM32CANType* const bxcan, const bool target_state)
+{
+    static const unsigned TimeoutMilliseconds = 2000;
+
+    for (unsigned wait_ack = 0; wait_ack < TimeoutMilliseconds; wait_ack++)
+    {
+        const bool state = (bxcan->MSR & CANARD_STM32_CAN_MSR_INAK) != 0;
+        if (state == target_state)
+        {
+            return true;
+        }
+
+        // Sleep 1 millisecond
+        usleep(1000);           // TODO: This function may be missing on some platforms
+    }
+
+    return false;
+}
+
+
+static void processErrorStatus(void)
+{
+    /*
+     * Aborting TX transmissions if abort on error was requested
+     * Updating error counter
+     */
+    const uint8_t lec = (uint8_t)((BXCAN->ESR & CANARD_STM32_CAN_ESR_LEC_MASK) >> CANARD_STM32_CAN_ESR_LEC_SHIFT);
+
+    if (lec != 0)
+    {
+        BXCAN->ESR = 0;                 // This action does only affect the LEC bits, other bits are read only!
+        g_stats.error_count++;
+
+        // Abort pending transmissions if auto abort on error is enabled, or if we're in bus off mode
+        if (g_abort_tx_on_error || (BXCAN->ESR & CANARD_STM32_CAN_ESR_BOFF))
+        {
+            BXCAN->TSR = CANARD_STM32_CAN_TSR_ABRQ0 | CANARD_STM32_CAN_TSR_ABRQ1 | CANARD_STM32_CAN_TSR_ABRQ2;
+        }
+    }
+}
+
+
+int canardSTM32Init(const CanardSTM32CANTimings* const timings,
+                    const CanardSTM32IfaceMode iface_mode)
```

Comment:
Nope, there's no way to do that in a completely OS and MCU agnostic way, unfortunately. I don't think this requirement will cause too much trouble to the user though, so it should be fine to leave it as is.